File: Test_suite/Java/graph.java
import java.util.*;

public class GraphDemo {

    static class Graph {
        private final Map<Integer, Map<Integer, Integer>> graph = new HashMap<>();

        public void addEdge(int from, int to, int weight) {
            graph.computeIfAbsent(from, k -> new HashMap<>()).put(to, weight);
            graph.computeIfAbsent(to, k -> new HashMap<>()).put(from, weight);
        }

        public Map<Integer, Integer> getNeighbors(int node) {
            return graph.getOrDefault(node, Collections.emptyMap());
        }
    }

    public static Graph generateLargeGraph(int numNodes, double edgeProbability, int maxWeight) {
        Graph graph = new Graph();
        Random random = new Random();

        for (int i = 0; i < numNodes; i++) {
            for (int j = i + 1; j < numNodes; j++) {
                if (random.nextDouble() < edgeProbability) {
                    int weight = random.nextInt(maxWeight) + 1;
                    graph.addEdge(i, j, weight);
                }
            }
        }
        return graph;
    }

    public static Map<Integer, Integer> dijkstra(Graph graph, int start) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        Map<Integer, Integer> distances = new HashMap<>();
        pq.offer(new int[]{start, 0});

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int currentNode = current[0];
            int currentDistance = current[1];

            if (!distances.containsKey(currentNode)) {
                distances.put(currentNode, currentDistance);

                for (Map.Entry<Integer, Integer> entry : graph.getNeighbors(currentNode).entrySet()) {
                    int nextNode = entry.getKey();
                    int nextDistance = currentDistance + entry.getValue();

                    if (!distances.containsKey(nextNode)) {
                        pq.offer(new int[]{nextNode, nextDistance});
                    }
                }
            }
        }
        return distances;
    }

    public static void main(String[] args) {
        int numNodes = 1000;
        double edgeProbability = 0.5;
        int maxWeight = 10;

        long startTime = System.currentTimeMillis();
        long beforeUsedMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory(); //memory calc start
        Graph graph = generateLargeGraph(numNodes, edgeProbability, maxWeight);
        Map<Integer, Integer> distances = dijkstra(graph, 0);
        long afterUsedMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory(); //memory calc end
        long actualMemUsed = afterUsedMem - beforeUsedMem; // Total memory took
        long actualMemUsedInMB = actualMemUsed / (1024 * 1024);
        long endTime = System.currentTimeMillis();
        

        System.out.println("Execution Time: " + (endTime - startTime) + " milliseconds");
        System.out.println("Memory Used: " + actualMemUsedInMB + " M");

        // Memory usage measurement is not straightforward in Java as in Python.
        // Java does not have a direct equivalent to Python's memory_profiler.
    }
}

File: Test_suite/Python/graph.py
import time
import heapq
import random
import resource

def generate_large_graph(num_nodes, edge_probability, max_weight=10):
    graph = {i: {} for i in range(num_nodes)}
    for i in range(num_nodes):
        for j in range(i + 1, num_nodes):
            if random.random() < edge_probability:
                weight = random.randint(1, max_weight)
                graph[i][j] = weight
                graph[j][i] = weight
    return graph

def dijkstra(graph, start):
    distances = {vertex: float('infinity') for vertex in graph}
    distances[start] = 0
    priority_queue = [(0, start)]

    while priority_queue:
        current_distance, current_vertex = heapq.heappop(priority_queue)

        if current_distance > distances[current_vertex]:
            continue

        for neighbor, weight in graph[current_vertex].items():
            distance = current_distance + weight

            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))

    return distances

def get_memory_usage():
    # Get the current memory usage in kilobytes
    memory_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    # Convert kilobytes to megabytes
    memory_mb = memory_kb / 1024.0
    return memory_mb

def main():
    num_nodes = 3000  # Number of nodes in the graph
    edge_probability = 0.5  # Probability of an edge between two nodes

    graph = generate_large_graph(num_nodes, edge_probability)

    start_node = 0
    start_time = time.time()
    distances = dijkstra(graph, start_node)
    end_time = time.time()

    print(f"Execution Time: {end_time - start_time} seconds")

if __name__ == "__main__":
    before_memory = get_memory_usage()
    main()
    after_memory = get_memory_usage()

    # Calculate peak memory usage during the execution
    peak_memory = max(before_memory, after_memory)
    print(f"Peak Memory Usage: {peak_memory} MiB")
